### PR TITLE
Add new --config option supporting config files

### PR DESCRIPTION
**Pull request by srawlins**

### CODE REVIEW

File: bin/sass_linter.dart [hunk 1]
```diff
@@ -23,17 +26,27 @@ void main(List<String> args) {
...
-        argResults.wasParsed('rules') ? _parseRules(argResults['rules']) : null;
+    // Generate a [Configuration], either by reading a configuration file, or
+    // by generating an empty one.
+    var config = argResults.wasParsed('config')
+        ? Configuration.parse(argResults['config'] as String)
+        : Configuration.empty;
+
+    var rules = argResults.wasParsed('rules')
+        ? parseRules(argResults['rules'] as List<String>)
+        : config.rules;
```
Comment (nex3) [inline bin/sass_linter.dart]: There may not be enough configuration for this to make sense now, but the API I ended up with in `test` was:

```dart
class Configuration {
  /// Parses configuration from command-line arguments.
  Configuration.fromArgs(List<String> arguments);

  /// Parses configuration from a path on disk.
  Configuration.load(String path);

  /// Returns a new configuration that has all the configuration defined in [this] or [other].
  ///
  /// Configuration in [other] takes precedence.
  Configuration merge(Configuration other);
}
```

Then all the configuration-parsing and -merging logic (including stuff like `parseRules`) lives in the `Configuration` class, and the executable can focus on business logic. It also makes it really easy to eventually do stuff like merging global and local configuration, if that becomes relevant.

File: lib/src/configuration.dart [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:yaml/yaml.dart';
+import 'package:sass_linter/src/engine.dart';
+import 'package:sass_linter/src/exceptions.dart';
+import 'package:sass_linter/src/rule.dart';
+
+class Configuration {
```
Comment (nex3) [inline lib/src/configuration.dart]: Nit: document this, and its members.

File: lib/src/configuration.dart [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  static final empty = Configuration._(null, <String>[]);
+
+  final List<Rule> rules;
+  final List<String> paths;
+
+  static YamlMap _loadYamlMap(String source, String path) {
+    var document = loadYamlNode(source, sourceUrl: p.toUri(path));
+
+    if (document is! Map) {
+      throw ConfigParseException('The configuration file must be a Yaml map.');
```
Comment (nex3) [inline lib/src/configuration.dart]: I strongly recommend making `ConfigParseException` a subclass of [`SourceSpanFormatException`](https://pub.dartlang.org/documentation/source_span/latest/source_span/SourceSpanFormatException-class.html) and passing in [`YamlNode.span`](https://pub.dartlang.org/documentation/yaml/latest/yaml/YamlNode/span.html). Then you can catch the exceptions and use [`SourceSpan.message()`](https://pub.dartlang.org/documentation/source_span/latest/source_span/SourceSpan/message.html) to print really nice-looking, colorful error messages.

File: test/cli_test.dart [hunk 4]
```diff
@@ -147,6 +147,101 @@ void main() {
...
+rules:
+  - unknown_rule
+''';
+    await d.file('a.yaml', configSource).create();
+    await d.file('a.scss', '@debug("here");').create();
+    var linter = await runLinter(['--config', '${d.sandbox}/a.yaml', 'a.scss']);
+
+    expect(linter.stdout, emits('Invalid rule: unknown_rule'));
+    await linter.shouldExit(64);
+  });
```
Comment (nex3) [inline test/cli_test.dart]: Add a test for a configuration that fails to parse.

File: lib/src/configuration.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  factory Configuration.parse(String path) {
+    var source = File(path).readAsStringSync();
+    if (source.isEmpty) return Configuration.empty;
+
+    var document = Configuration._loadYamlMap(source, path);
+    if (document.value == null) return Configuration.empty;
+
+    List<Rule> rules;
+
+    if (document.containsKey('rules')) {
```
Comment (nex3) [inline lib/src/configuration.dart]: I generally like allowing `null` rules to have the default values as well, just to be a little more flexible for the user.

File: lib/src/configuration.dart [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      if (document['rules'] is! List) {
+        throw ConfigParseException(
+            'The value of "rules" in the configuration file is expected to be '
+            'a List.');
+      }
+      rules = List<Rule>.unmodifiable(
+          parseRules((document['rules'] as List).cast<String>()));
```
Comment (nex3) [inline lib/src/configuration.dart]: `.cast<String>` won't throw a very friendly error if the user accidentally includes a non-string. At the risk of making you sick of `test` references, [this is what I did there](https://github.com/dart-lang/test/blob/master/pkgs/test_core/lib/src/runner/configuration/load.dart#L463-L471).

Also, make sure a rule that doesn't exist throws a nice error here.

File: test/configuration_test.dart [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    await d.file('a.yaml', configSource).create();
+    var configuration = Configuration.parse('${d.sandbox}/a.yaml');
+    expect(configuration.rules, hasLength(2));
+    expect(configuration.rules, contains(TypeMatcher<NoDebugRule>()));
+    expect(configuration.rules, contains(TypeMatcher<NoLoudCommentRule>()));
+    expect(configuration.paths, hasLength(2));
+    expect(configuration.paths, contains('foo/bar.scss'));
+    expect(configuration.paths, contains('foo/baz.scss'));
+  });
+}
```
Comment (nex3) [inline test/configuration_test.dart]: It's probably a good idea to test all the error cases here, too. I know it's kind of tedious, but I've definitely caught bugs from tests like that.

File: lib/src/configuration.dart (hunk 3 above)
Comment (srawlins) [reply lib/src/configuration.dart]: Great suggestion! Done.

File: bin/sass_linter.dart [hunk 8]
```diff
@@ -23,20 +26,33 @@ void main(List<String> args) {
...
+
+      if (paths.isEmpty) _usage('Report lint found in Sass');
 
-      var engine = new Engine(argResults.rest,
+      var engine = new Engine(paths,
           rules: rules, stdinFileUrl: argResults['stdin-file-url'] as String);
       _report(engine.run());
     }
+  } on ConfigParseException catch (error) {
+    print(error.toString(color: '\u001b[31m'));
```
Comment (nex3) [inline bin/sass_linter.dart]: Red is already the default color, so you can just pass `color: true` here. Or better, detect whether the user's terminal is likely to support color codes (see [the Dart Sass](https://github.com/sass/dart-sass/blob/master/lib/src/io/vm.dart#L26-L33) implementation for example).

File: lib/src/configuration.dart (hunk 5 above)
Comment (nex3) [reply lib/src/configuration.dart]: Yeah, exactly. In this case, `document['rules']` will be `null`.

File: lib/src/configuration.dart (hunk 6 above)
Comment (nex3) [reply lib/src/configuration.dart]: Yeah... at one point I had an idea for a nice generic API for extracting typed values from untyped objects with useful error messages, but when I proposed it it just descended into bickering so I never did it :woman_shrugging:.

File: bin/sass_linter.dart [hunk 9]
```diff
@@ -58,17 +75,9 @@ void _report(Iterable<Lint> lints) {
...
-Iterable<Rule> _parseRules(List<String> rules) => rules.map((ruleName) {
-      var sanitizedName =
-          ruleName.endsWith('_rule') ? ruleName : '${ruleName}_rule';
-      try {
-        return allRules.firstWhere((r) => r.name == sanitizedName);
-      } on StateError {
-        throw new UsageException('Invalid rule: $ruleName');
-      }
-    });
+bool get _supportsAnsiEscapes =>
```
Comment (nex3) [inline bin/sass_linter.dart]: Nit: document this.

File: lib/src/configuration.dart [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    return document as YamlMap;
+  }
+
+  /// Asserts that [field] is a list and runs [forElement] for each element it
+  /// contains.
+  ///
+  /// Returns a list of values returned by [forElement].
+  static List<T> _getList<T>(
+      YamlMap document, String field, T forElement(YamlNode elementNode)) {
+    var node = document.nodes[field]; // (value) => value is List) as YamlList;
```
Comment (nex3) [inline lib/src/configuration.dart]: What's up with this comment?

File: test/configuration_test.dart (hunk 7 above)
Comment (nex3) [reply test/configuration_test.dart]: I strongly recommend including tests for *every* type of error that can come from an invalid file. It's really easy for individual errors to sneak through in ways that make the program fail ungracefully. For example, does it throw a `ConfigParseException` if a rule name doesn't exist?

File: bin/sass_linter.dart (hunk 9 above)
Comment (srawlins) [reply bin/sass_linter.dart]: Done.

File: lib/src/configuration.dart (hunk 10 above)
Comment (srawlins) [reply lib/src/configuration.dart]: My bad. Half-commented out code.

File: test/configuration_test.dart (hunk 7 above)
Comment (srawlins) [reply test/configuration_test.dart]: Most of the tests I added are in cli_test.dart, to assert on the message given to the user and the exit code. But I can assert on the exception type here as well.

Comment (srawlins) [reply test/configuration_test.dart]: Added a few more in configuration_test too.

File: test/configuration_test.dart [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+paths:
+''';
+    await d.file('a.yaml', configSource).create();
+    var configuration = Configuration.parse('${d.sandbox}/a.yaml');
+    expect(configuration.rules, isNull);
+    expect(configuration.paths, isEmpty);
+  });
+
+  test('throws when a configuration file cannot be parsed as YAML', () async {
+    await _assertError<YamlException>(r'''
```
Comment (nex3) [inline test/configuration_test.dart]: This should probably be converted to a `ConfigParseException` for consistency. Callers of your parser shouldn't have to know or care about the underlying format the config file uses, or its API.

File: test/configuration_test.dart [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+  test('throws when a configuration file cannot be parsed as a map', () async {
+    await _assertError<ConfigParseException>(r'''
+- no_debug_rule
+- no_loud_comment_rule
+''');
+  });
+
+  test('throws when a configuration file cannot be parsed', () async {
+    await _assertError<UsageException>(r'''
```
Comment (nex3) [inline test/configuration_test.dart]: This should definitely be a `ConfigParseException`. A `UsageException` won't highlight the specific rule, and may not even indicate that it comes from the config file.

Comment (srawlins) [reply test/configuration_test.dart]: OK, I've reconfigured the engine to throw an UnknownRuleException, which can either be handled as a UsageException, when passed on the command line, or a ConfigParseException, when found in a config.

File: test/configuration_test.dart (hunk 11 above)
Comment (srawlins) [reply test/configuration_test.dart]: Done.

**Comment by srawlins** [comment]
Woohoo, thanks!

**Pull request merged**